Draft: feat: add Intl.DurationFormat polyfill implementation

ewlsh opened the pull request:
This MR adds a to-spec implementation of the Intl.DurationFormat polyfill implementation. I'm still working on microsecond and nanosecond support in Intl.NumberFormat, something is breaking when running `test262/test/intl402/DurationFormat/prototype/format/style-options-en.js` - the test passes with the below modification...

```
const duration = {
  years: 1,
  months: 2,
  weeks: 3,
  days: 3,
  hours: 4,
  minutes: 5,
  seconds: 6,
  milliseconds: 7,
  // microseconds: 8,
  // nanoseconds: 9,
};
```

MR is a WIP, will be updated with more details. Any initial feedback is helpful, this is my first time working on this codebase.

ewlsh commented on packages/intl-localematcher/abstract/LookupSupportedLocales.ts:
```diff
@@ -21,7 +21,8 @@ export function LookupSupportedLocales(
     )
     if (availableLocale) {
-      subset.push(availableLocale)
+      // test262/test/intl402/DurationFormat/supportedLocalesOf/locales-specific.js
```
This seems like a pre-existing bug, we're returning the resolved locales not the inputted ones.

ewlsh commented on packages/ecma402-abstract/types/duration.ts:
```diff
@@ -0,0 +1,10 @@
+import {LocaleData} from './core'
+
+export interface DurationLocaleData {
+  digitalFormat: {
```
Currently only the default numbering system is supported in a locale, this could be converted to something like `Record<string, { digitalFormat: { separator: string } }>` to store the information per-number system.

ewlsh commented on packages/intl-listformat/index.ts:
```diff
@@ -149,13 +157,33 @@ function createPartsFromList(
   return parts
 }
 
+export function CreatePartsFromList(
```
TODO: Add doc link here

ewlsh commented on packages/intl-locale/index.ts:
```diff
@@ -41,8 +42,6 @@ export interface IntlLocaleInternal extends IntlLocaleOptions {
   initializedLocale: boolean
 }
 
-const UNICODE_TYPE_REGEX = /^[a-z0-9]{3,8}(-[a-z0-9]{3,8})*$/i
```
Abstracting this operation so duration format can also use it...

ewlsh commented on packages/intl-listformat/index.ts:
```diff
@@ -119,24 +119,32 @@ function stringListFromIterable(list: any[]): string[] {
   internalSlotMap: WeakMap<ListFormat, ListFormatInternal>,
   lf: ListFormat,
-  list: string[]
+  list: (string | {type: string; value: string})[]
```
This change is listed in the durationformat spec

ewlsh commented on packages/intl-durationformat/test262-main.ts:
```diff
@@ -0,0 +1,1157 @@
+import '@formatjs/intl-locale/polyfill'
+import '@formatjs/intl-enumerator/polyfill'
+
+import '@formatjs/intl-listformat/polyfill-force'
```
DurationFormat relies on polyfilled changes to ListFormat and NumberFormat... I didn't see the other test262 loading polyfills so if there is another way to approach this I can alter this.

ewlsh commented on packages/intl-durationformat/BUILD:
```diff
@@ -1,21 +1,40 @@
 PACKAGE_NAME = "intl-durationformat"
 
+
+TEST_LOCALES = [
```
Just copied from ListFormat, I haven't put time into building actually useful unit tests for DurationFormat yet - have been focused on test262

longlho commented:
it's very unlikely all the test262 cases are gonna pass, especially w/ tests that involves `Realm`, so you might wanna exclude some

ewlsh commented:
> it's very unlikely all the test262 cases are gonna pass, especially w/ tests that involves `Realm`, so you might wanna exclude some

@longlho currently I have all DurationFormat cases passing locally except the test listed above - I don't see anything Realm related in the current DurationFormat test suite here: https://github.com/tc39/test262/tree/main/test/intl402/DurationFormat

~I did have to update the git submodule due to changes in the spec/test suite recently.~ I reverted because it wasn't playing well with CI, I rebased on `main` so it is possible the version on `main` is new enough to have the updated DurationFormat tests.

```
bazel test //packages/intl-durationformat:test262
```

currently yields

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //packages/intl-durationformat:test262
-----------------------------------------------------------------------------

PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-locales-invalid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-locales-invalid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-locales-valid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-locales-valid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-defaults.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-defaults.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-fractionalDigits-invalid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-fractionalDigits-invalid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-fractionalDigits-valid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-fractionalDigits-valid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-invalid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-invalid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-localeMatcher-invalid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-localeMatcher-invalid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-localeMatcher-valid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-localeMatcher-valid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-numberingSystem-invalid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-numberingSystem-invalid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-numberingSystem-valid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-numberingSystem-valid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-order.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-order.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-style-invalid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-style-invalid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-style-valid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/constructor-options-style-valid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/extensibility.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/extensibility.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/length.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/length.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/name.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/name.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/newtarget-undefined.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/newtarget-undefined.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prop-desc.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prop-desc.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/prototype_attributes.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/prototype_attributes.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/supportedLocalesOf/basic.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/supportedLocalesOf/basic.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/supportedLocalesOf/branding.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/supportedLocalesOf/branding.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/supportedLocalesOf/length.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/supportedLocalesOf/length.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/supportedLocalesOf/locales-empty.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/supportedLocalesOf/locales-empty.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/supportedLocalesOf/locales-invalid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/supportedLocalesOf/locales-invalid.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/supportedLocalesOf/locales-specific.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/supportedLocalesOf/locales-specific.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/supportedLocalesOf/name.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/supportedLocalesOf/name.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/supportedLocalesOf/prop-desc.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/supportedLocalesOf/prop-desc.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/constructor/prop-desc.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/constructor/prop-desc.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/constructor/value.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/constructor/value.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/basic-format-en.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/basic-format-en.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/branding.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/branding.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/invalid-arguments-throws.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/invalid-arguments-throws.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/invalid-negative-duration-throws.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/invalid-negative-duration-throws.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/length.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/length.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/name.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/name.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/not-a-constructor.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/not-a-constructor.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/prop-desc.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/prop-desc.js
FAIL ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/style-options-en.js (default)
  Expected no error, got TypeError: Cannot read properties of undefined (reading 'en')

FAIL ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/style-options-en.js (strict mode)
  Expected no error, got TypeError: Cannot read properties of undefined (reading 'en')

PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/throw-invoked-as-func.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/format/throw-invoked-as-func.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/formatToParts/branding.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/formatToParts/branding.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/formatToParts/invalid-arguments-throws.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/formatToParts/invalid-arguments-throws.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/formatToParts/invalid-negative-duration-throws.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/formatToParts/invalid-negative-duration-throws.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/formatToParts/length.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/formatToParts/length.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/formatToParts/name.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/formatToParts/name.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/formatToParts/not-a-constructor.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/formatToParts/not-a-constructor.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/formatToParts/prop-desc.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/formatToParts/prop-desc.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/formatToParts/throw-invoked-as-func.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/formatToParts/throw-invoked-as-func.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/resolvedOptions/length.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/resolvedOptions/length.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/resolvedOptions/name.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/resolvedOptions/name.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/resolvedOptions/prop-desc.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/resolvedOptions/prop-desc.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/resolvedOptions/throw-invoked-as-func.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/resolvedOptions/throw-invoked-as-func.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/toStringTag/toString.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/toStringTag/toString.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/toStringTag/toStringTag.js
PASS ../com_github_tc39_test262/test/intl402/DurationFormat/prototype/toStringTag/toStringTag.js
Ran 108 tests
106 passed
2 failed

```

ewlsh committed:
feat: add Intl.DurationFormat polyfill implementation

longlho commented:
I just pushed a fix for test262 initialization so you should rebase

ewlsh commented:
> I just pushed a fix for test262 initialization so you should rebase

I've rebased and re-generated the test data. The action/CI pipeline to build it doesn't seem to be triggering though.

longlho commented:
is this ready for review?

romulocintra commented:
> is this ready for review?

@ewlsh and @longlho  do you need want some help on this ? Would be interesting having a polyfill implementation, can we try to work on this in case we have some cycles ? cc @sffc @ben-allen @ryzokuken

longlho commented:
looks like there're a buncha TODOs in the diff so wanna make sure it's ready for review before I review it

ewlsh commented:
> looks like there're a buncha TODOs in the diff so wanna make sure it's ready for review before I review it

I believe there are 3 TODOs added (and I just resolved one)...
* One is from a build config I copied (`# TODO: fix`)
* One is for the karma tests (` # TODO: "//packages/intl-durationformat:bundled-karma-tests",`) - I haven't sorted how these are supposed to work if I remember correctly.
* The last one (which is now fixed) was to add a documentation link to the pre-existing utility function in listformat

I removed the TODOs that were placeholders in this package.

An initial review would be helpful to move this forward, this is my first attempt using bazel and proposing changes to this repo.

> something is breaking when running test262/test/intl402/DurationFormat/prototype/format/style-options-en.js - the test passes with the below modification...

If you have any advice on this issue, I was struggling to figure out what was breaking in the current numberformat implementation, I believe I updated it to support the new units but something was still not working as expected :thinking: (I'll also have a look again and see if I can sort out what is going wrong)

ewlsh commented:
> > is this ready for review?
> 
> @ewlsh and @longlho do you need want some help on this ? Would be interesting having a polyfill implementation, can we try to work on this in case we have some cycles ? cc @sffc @ben-allen @ryzokuken

I think this is fairly close, the main issues I believe are...

* Support for non-default number systems (if we want this)
* Fixing up remaining test cases
* Figuring out issue with numberformat polyfill changes

Any help is appreciated :smile:

longlho commented:
ok I'm ok w/ merging this since it's a pretty big PR. Can u create GH issues for the TODOs & make sure build passes at least locally? Some test262 tests are kinda impossible to pass (esp the `Realm` stuff) so I'm ok w/ it not passing 100%.
